chiselFrontend/src/main/scala/chisel3/internal/Builder.scala

```diff
@@ -223,6 +225,36 @@ private[chisel3] object Builder {
     pushCommand(cmd).id
   }
 
+  // Called when Bundle construction begins, used to record a stack of open Bundle constructors to
+  // record candidates for Bundle autoclonetype. This is a best-effort guess.
+  // Returns the current stack of open Bundles
+  // Note: elt will NOT have finished construction, its elements cannot be accessed
+  def updateBundleStack(elt: Bundle): Seq[Bundle] = {
+    val stackClasses = Thread.currentThread().getStackTrace()
+        .map(_.getClassName)
+        .reverse  // so stack frame numbers are deterministic across calls
+
+    // Prune the existing Bundle stack of closed Bundles
+    while (!dynamicContext.bundleStack.isEmpty &&
+        (dynamicContext.bundleStack.last._3 >= stackClasses.size ||
+            (stackClasses(dynamicContext.bundleStack.last._3) != dynamicContext.bundleStack.last._2))) {
+      dynamicContext.bundleStack.remove(dynamicContext.bundleStack.size - 1)
+    }
```

Comment:
I think this could be better expressed as something like:
```scala
while ({
  val (_, cname, pos) = dynamicContext.bundleStack.last
  !dynamicContext.bundleStack.isEmpty && pos >= stackClasses.size || stackClasses(pos) != cname
}) {
  dynamicContext.bundleStack.trimEnd(1)
}

Reply:
A block in a while condition, while probably legal Scala, seems like quite odd style.
Anyways, I restructured the whole thing to sidestep the issue, it first calculates the length to prune using prefixLength, then trims that.